src/namedtuples.jl

```diff
@@ -1,54 +1,112 @@
+        return quote
+            rows.st === nothing && return nothing
+            row, st = rows.st
+            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), st
+        end
+    else
+        rows.st === nothing && return nothing
+        row, st = rows.st
+        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getcolumn(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), st
+    end
+end
+
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, T}}, st) where {names, T}
     if @generated
-        vals = Tuple(:(getproperty(row, $(fieldtype(T, i)), $i, $(quot(names[i])))) for i = 1:fieldcount(T))
+        vals = Tuple(:(getcolumn(row, $(fieldtype(T, i)), $i, $(quot(names[i])))) for i = 1:fieldcount(T))
         return quote
-            x = iterate(rows.x, st...)
+            x = iterate(rows.x, st)
             x === nothing && return nothing
             row, st = x
-            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), (st,)
+            return $(NamedTuple{Base.map(Symbol, names), T})(($(vals...),)), st
         end
     else
-        x = iterate(rows.x, st...)
+        x = iterate(rows.x, st)
         x === nothing && return nothing
         row, st = x
-        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getproperty(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), (st,)
+        return NamedTuple{Base.map(Symbol, names), T}(Tuple(getcolumn(row, fieldtype(T, i), i, names[i]) for i = 1:fieldcount(T))), st
     end
 end
 
-# unknown schema case
-function Base.iterate(rows::NamedTupleIterator{Nothing, T}, st=()) where {T}
-    x = iterate(rows.x, st...)
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, nothing}}) where {names}
+    rows.st === nothing && return nothing
+    row, st = rows.st
+    return NamedTuple{names}(Tuple(getcolumn(row, nm) for nm in names)), st
+end
+
+@inline function Base.iterate(rows::NamedTupleIterator{Schema{names, nothing}}, st) where {names}
+    x = iterate(rows.x, st)
     x === nothing && return nothing
     row, st = x
-    names = Tuple(propertynames(row))
-    return NamedTuple{Base.map(Symbol, names)}(Tuple(getproperty(row, nm) for nm in names)), (st,)
+    return NamedTuple{names}(Tuple(getcolumn(row, nm) for nm in names)), st
 end
 
-namedtupleiterator(::Type{T}, rows::S) where {T <: NamedTuple, S} = rows
-namedtupleiterator(::Type{T}, rows::S) where {T, S} = NamedTupleIterator{typeof(schema(rows)), S}(rows)
-
 # sink function
+"""
+    Tables.rowtable(x) => Vector{NamedTuple}
+    Tables.rowtable(rt, x) => rt
+
+Take any input table source, and produce a `Vector` of `NamedTuple`s,
+also known as a "row table". A "row table" is a kind of default
+table type of sorts, since it satisfies the Tables.jl row interface
+naturally.
+
+The 2nd definition takes
+an existing row table and appends the input table source `x`
+to the existing row table.
+"""
+function rowtable end
```

Comment:
I suggest to add `isrowtable` trait function #134.  It would introduce "row table" to mean something different.  I'm not sure if you want to include #134 in 1.0.  But can you at least remove `Tables.rowtable` from the public API?  Of course, rejecting #134 and keeping `Tables.rowtable` is a reasonable choice, too.